src/cli/zstore.rs

```diff
@@ -135,21 +136,65 @@ impl<F: LurkField> ZDag<F> {
                 z_ptr
             }
         };
-        recurse(ptr)
+        let mut ptrs: IndexSet<Ptr> = IndexSet::default();
+        let mut stack = vec![*ptr];
+        macro_rules! feed_loop {
+            ($x:expr) => {
+                if !cache.contains_key(&$x) {
+                    if ptrs.insert($x) {
+                        stack.push($x);
+                    }
+                }
+            };
+        }
+        while let Some(ptr) = stack.pop() {
+            match ptr.raw() {
+                RawPtr::Atom(..) => (),
+                RawPtr::Hash4(idx) => {
+                    for ptr in expect_ptrs!(store, 2, *idx) {
+                        feed_loop!(ptr)
+                    }
+                }
+                RawPtr::Hash6(idx) => {
+                    for ptr in expect_ptrs!(store, 3, *idx) {
+                        feed_loop!(ptr)
+                    }
+                }
+                RawPtr::Hash8(idx) => {
+                    for ptr in expect_ptrs!(store, 4, *idx) {
+                        feed_loop!(ptr)
+                    }
+                }
+            }
+        }
+        ptrs.iter().rev().for_each(|ptr| {
+            recurse(ptr, cache);
+        });
+        recurse(ptr, cache)
```

Comment:
Why not add `ptr` to the initial value of `ptrs` instead? It would make this line redundant, no?

Reply:
Here's the suggested diff:
```
diff --git a/src/cli/zstore.rs b/src/cli/zstore.rs
index 892fa97a4..2be2e6f61 100644
--- a/src/cli/zstore.rs
+++ b/src/cli/zstore.rs
@@ -136,7 +136,7 @@ impl<F: LurkField> ZDag<F> {
                 z_ptr
             }
         };
-        let mut ptrs: IndexSet<Ptr> = IndexSet::default();
+        let mut ptrs: IndexSet<Ptr> = IndexSet::from([*ptr]);
         let mut stack = vec![*ptr];
         macro_rules! feed_loop {
             ($x:expr) => {
@@ -167,10 +167,11 @@ impl<F: LurkField> ZDag<F> {
                 }
             }
         }
+        let mut res = None;
         ptrs.iter().rev().for_each(|ptr| {
-            recurse(ptr, cache);
+            res = Some(recurse(ptr, cache));
         });
-        recurse(ptr, cache)
+        res.unwrap()
     }
     ```
     If it violates correctness, can you please add a test that breaks with this snippet, and passes without? It would be an super clean way to make a correctness condition that is not at the moment clear to me.

Reply:
Yeah, conceptually it's the same thing (we call `recurse` the same number of times). But as I said, the code gets a little worse with this mutable `Option` that needs to be unwrapped in the end. Plus the assumption I mentioned about that `ptr` will be in the last iteration of the reverse iterator

Reply:
The correctness condition is in `test_z_store_roundtrip` by the way

Reply:
`test_z_store_roudntrip` passes in both cases. You can also get away with the following:
```rust
        ptrs.iter().rev().map(|ptr| {
            recurse(ptr, cache)
        }).last().unwrap();
```
It's just a nit, but this pattern is repeated 3 times ..